### ISSUE TITLE
[Spike] Gather the recent locations for a given user in the Appointments tool on VA.gov

### ISSUE THREAD

**Issue by ldelacosta** [department-of-veterans-affairs/va.gov-team#81136]
## Background/Request
We are looking to improve the user experience for the locations page in the Appointments tool by displaying the most recent locations for a user. 

### Goal
The user will be able to see a list of the recent VA locations they have visited or will visit in the future. 

### Requirements to Consider
<img width="171" alt="Screenshot 2024-04-19 at 11 26 23 AM" src="https://github.com/department-of-veterans-affairs/va.gov-team/assets/90797205/65f563d0-9ccc-4c2b-a158-fc0d04af88bd">

---

## Tasks
- Answer the following questions:
  - Q1: Can we create an endpoint to provide a way to filter facility location by recently visited? (looking at upcoming appointments and past appointments) 
  - Q2: What's the best approach for this?
  - Q3: what's the performance hit of checking a facilities recently visited in the last 36 months

## Time Box
8 hours

## Definition of Done
- Tasks are complete
- Research and recommendations are documented in this ticket

---

**Comment by ajmagdub** [comment]
Under ticket [52135] (https://github.com/department-of-veterans-affairs/va.gov-team/issues/52135), an endpoint to get the most recently visited clinic was created. The Vets API endpoint is **/v2/locations/last_visited_clinic**.

Logic for retrieving the last visited clinic is in the`last_visited_clinic` in the `VAOS::V2::ClinicsController#last_visited_clinic` action.

The method begins by calling the `get_most_recent_visited_clinic_appointment` method on the `appointments_service` object. This retrieves the most recent clinic appointment for the current user. The result of this call is stored in the `latest_appointment` variable.

```ruby
latest_appointment = appointments_service.get_most_recent_visited_clinic_appointment
```

Next, the method checks if it's possible to look up the clinic based on the information in `latest_appointment`. This is done by calling the `unable_to_lookup_clinic?` method with `latest_appointment` as the argument. If it's not possible to look up the clinic (i.e., the method returns `true`), the method logs this issue by calling `log_unable_to_lookup_clinic(latest_appointment)`, sends a `404` status code with a message, and then exits.

```ruby
if unable_to_lookup_clinic?(latest_appointment)
  log_unable_to_lookup_clinic(latest_appointment)
  render json: { message: 'Unable to lookup latest clinic.' }, status: :not_found
  return
end
```

If it is possible to look up the clinic, the method proceeds to retrieve the clinic details. It does this by calling the `get_clinic_with_cache` method on the `mobile_facility_service` object, passing in the `station_id` and `clinic_id` from `latest_appointment`.

```ruby
station_id = latest_appointment.location_id
clinic_id = latest_appointment.clinic
clinic = mobile_facility_service.get_clinic_with_cache(station_id:, clinic_id:)
```

The method then checks if the clinic details were successfully retrieved. If not (i.e., `clinic` is `nil`), it logs this issue, sends a `404` status code with a message, and then exits.

```ruby
if clinic.nil?
  log_no_clinic_details_found(station_id, clinic_id)
  render json: { message: 'No clinic details found' }, status: :not_found
  return
end
```

Finally, if the clinic details were successfully retrieved, the method returns these details as JSON by rendering them with the `VAOS::V2::ClinicsSerializer`.

```ruby
render json: VAOS::V2::ClinicsSerializer.new(clinic)
```


The helper method `VAOS::V2::AppointmentsService#get_most_recent_visited_clinic_appointment`, is designed to retrieve the most recent appointment for a patient within the last year. 

The method starts by setting up a few variables. `current_check` is set to the end of the previous day. `three_month_interval` is set to a duration of three months, and `look_back_limit` is set to one year ago from the current date. The `statuses` variable is set to a string containing the statuses 'booked', 'fulfilled', and 'arrived'.

```ruby
current_check = Date.current.end_of_day.yesterday
three_month_interval = 3.months
look_back_limit = 1.year.ago
statuses = 'booked,fulfilled,arrived'
```

The method then enters a loop that continues until `current_check` is earlier than `look_back_limit`. Inside the loop, it sets `end_time` to the current check date and `start_time` to three months before the current check date. 

```ruby
end_time = current_check
start_time = current_check - three_month_interval
```

It then calls the `fetch_clinic_appointments` method with `start_time`, `end_time`, and `statuses` as arguments. This method is expected to return a list of clinic appointments within the specified time range and with the specified statuses.

```ruby
appointments = fetch_clinic_appointments(start_time, end_time, statuses)
```

If the `appointments` list is not empty, the method returns the most recent appointment by calling the `most_recent_appointment` method with `appointments` as the argument. If the `appointments` list is empty, it subtracts `three_month_interval` from `current_check` and continues to the next iteration of the loop.

```ruby
return most_recent_appointment(appointments) unless appointments.empty?
current_check -= three_month_interval
```

If the loop completes without finding any appointments (i.e., `current_check` becomes earlier than `look_back_limit`), the method returns `nil`. This indicates that no clinic appointments were found within the last year.



 In summary, the `last_visited_clinic` method retrieves the most recent clinic appointment for the current user, checks if it's possible to look up the clinic based on this appointment, retrieves the clinic details if possible, and then returns these details as JSON. If at any point something goes wrong (i.e., it's not possible to look up the clinic or the clinic details can't be retrieved), the method logs the issue and sends a `404` status code with a message.

**Comment by ldelacosta** [comment]
@melissa-lefevre

**Comment by ldelacosta** [comment]
Added the spike ticket to the [feature doc](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/health-care/appointments/va-online-scheduling/initiatives/scheduling-improvements/scheduling-improvements.md). And closing out the ticket.
